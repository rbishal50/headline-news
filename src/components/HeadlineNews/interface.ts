export interface IProps {
  clicked: (data: INews) => void
}

export interface INews {
  id?: string
  title: string
  subtitle: string
  content: string
  imageUrl: string
}
