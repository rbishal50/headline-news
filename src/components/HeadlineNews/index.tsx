import { FC, useEffect } from 'react'
import PropTypes from 'prop-types'
import { Row, Col } from 'react-bootstrap'
import { useSelector, useDispatch, shallowEqual } from 'react-redux'

import { RootState } from '../../redux/rootReducer'
import { fetchHeadlines } from '../../redux/slices/newsSlice'
import ListItems from '../ListItems'
import PrimaryCard from '../PrimaryCard'
import Loader from '../Spinner'
import { IProps } from './interface'

export const HeadlineNews: FC<IProps> = ({ clicked }) => {
  const dispatch = useDispatch()

  const { headlines, isLoading, error } = useSelector((state: RootState) => {
    return {
      headlines: state.news.headlines,
      isLoading: state.news.isLoading,
      error: state.news.error,
    }
  }, shallowEqual)

  useEffect(() => {
    dispatch(fetchHeadlines())
  }, [dispatch])

  const handleCardDisplay = (index: number, data: any) => {
    if (index === 0) {
      return (
        <Col md='12' key={data.id}>
          <PrimaryCard
            title={data.title}
            clicked={clicked}
            type='imageToTheRight'
            content={data.content}
            subtitle={data.subtitle}
            imageUrl={data.imageUrl}
          />
        </Col>
      )
    } else {
      return (
        <Col md={index === 1 ? '8' : '4'} key={index}>
          <PrimaryCard
            title={data.id}
            clicked={clicked}
            type='normal'
            content={index === 1 ? data.content : ''}
            subtitle={data.subtitle}
            size={index === 1 ? 'lg' : 'sm'}
            imageUrl={data.imageUrl}
          />
          {index > 1 && <ListItems items={headlines.slice(3, 6)} clicked={clicked} />}
        </Col>
      )
    }
  }

  return (
    <>
      {isLoading && (
        <div data-testid='loader'>
          <Loader />
        </div>
      )}
      {error && <p data-testid='error'>{error}</p>}
      <Row>
        {!isLoading &&
          (headlines.slice(0, 3) || []).map((headline: any, index: number) => handleCardDisplay(index, headline))}
      </Row>
    </>
  )
}

HeadlineNews.propTypes = {
  clicked: PropTypes.func.isRequired,
}

export default HeadlineNews
