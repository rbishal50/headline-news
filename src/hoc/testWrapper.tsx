import { ComponentType, ReactElement } from 'react'
import { BrowserRouter } from 'react-router-dom'
import { Provider } from 'react-redux'

import store from '../redux/store'

const testWrapper = (Component: any, props = {}): ReactElement => {
  return (
    <BrowserRouter>
      <Provider store={store}>
        <Component {...props} />
      </Provider>
    </BrowserRouter>
  )
}

export default testWrapper
