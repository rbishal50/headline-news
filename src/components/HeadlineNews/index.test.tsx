import { render, screen } from '@testing-library/react'
import * as reactRedux from 'react-redux'

import testWrapper from '../../hoc/testWrapper'
import HeadlineNews from './index'

const mockFunction = jest.fn()

describe('<HeadlineNews />', () => {
  const useSelectorMock = jest.spyOn(reactRedux, 'useSelector')

  beforeEach(() => {
    useSelectorMock.mockClear()
  })

  it('renders items', () => {
    useSelectorMock.mockReturnValue({
      headlines: [
        { id: '11', title: 'test title 1' },
        { id: '22', title: 'test title 2', subtitle: 'subtitle 2' },
      ],
    })
    const { container } = render(testWrapper(HeadlineNews, { clicked: mockFunction }))
    expect(container).toMatchSnapshot()
  })

  it('renders a spinner when loading', () => {
    useSelectorMock.mockReturnValue({
      isLoading: true,
    })
    render(testWrapper(HeadlineNews, { clicked: mockFunction }))
    screen.getByTestId('loader')
  })

  it('renders error message if present', () => {
    useSelectorMock.mockReturnValue({
      headlines: [],
      error: 'Error',
    })
    render(testWrapper(HeadlineNews, { clicked: mockFunction }))
    screen.getByTestId('error')
  })
})
