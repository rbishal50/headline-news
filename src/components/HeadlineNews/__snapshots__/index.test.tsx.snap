// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HeadlineNews /> renders a spinner when loading is true 1`] = `
<div>
  <div
    class="loader"
  >
    <div
      class="spinner-grow text-danger"
    />
  </div>
  <div
    class="row"
  />
</div>
`;

exports[`<HeadlineNews /> renders items 1`] = `
<div>
  <div
    class="row"
  >
    <div
      class="col-md-12"
    >
      <div
        class="primary-card mb-4 flex-row-reverse card"
        data-testid="card"
      >
        <div>
          <img
            alt="Cover"
            class="card-img cover"
            src="https://picsum.photos/500/300"
          />
        </div>
        <div>
          <div
            class="card-body"
            style="max-width: 500px;"
          >
            <div
              class="card-title card-title h5"
              style="font-size: 30px;"
            >
              test title 1
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="col-md-8"
    >
      <div
        class="primary-card mb-4 flex-column card"
        data-testid="card"
      >
        <div>
          <img
            alt="Cover"
            class="card-img cover"
            src="https://picsum.photos/500/300"
          />
        </div>
        <div>
          <div
            class="card-body"
            style="max-width: 500px;"
          >
            <div
              class="card-title card-title h5"
              style="font-size: 30px;"
            >
              22
            </div>
            <p
              class="text-muted subtitle"
            >
              subtitle 2
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<HeadlineNews /> renders items 2`] = `
<div>
  <div
    class="row"
  >
    <div
      class="col-md-12"
    >
      <div
        class="primary-card mb-4 flex-row-reverse card"
        data-testid="card"
      >
        <div>
          <img
            alt="Cover"
            class="card-img cover"
            src="https://picsum.photos/500/300"
          />
        </div>
        <div>
          <div
            class="card-body"
            style="max-width: 500px;"
          >
            <div
              class="card-title card-title h5"
              style="font-size: 30px;"
            >
              test title 1
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="col-md-8"
    >
      <div
        class="primary-card mb-4 flex-column card"
        data-testid="card"
      >
        <div>
          <img
            alt="Cover"
            class="card-img cover"
            src="https://picsum.photos/500/300"
          />
        </div>
        <div>
          <div
            class="card-body"
            style="max-width: 500px;"
          >
            <div
              class="card-title card-title h5"
              style="font-size: 30px;"
            >
              22
            </div>
            <p
              class="text-muted subtitle"
            >
              subtitle 2
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
